whole-tale/girder_wholetale
# Add new rest endpoint for PUT /instance/:id

# Problem
Fixes #207  - once a Tale has been rebuilt, we need a way for the user to signal to the underlying instance that it should be restarted with the newest image.

Relates to https://github.com/whole-tale/gwvolman/pull/44

# Approach
Add a general `update_container` task that can be easily expanded to cover additional related functionality, as needed.

# How to Test
There is currently not a super clean way to test this, as it currently requires running this code alongside the code from another PR to `girder_wholetale`.

So here goes:
1. Run [`deploy-dev`](https://github.com/whole-tale/deploy-dev/tree/dev) stack using the provided README instructions (make sure to use `dev` branch for now)
2. `cd src/gwvolman` and checkout the `add_new_task_update_container` branch
    * NOTE: You may need to restart the worker with `make restart_worker`
3. `cd ../../src/wholetale` and checkout this branch
    * NOTE: Girder will restart automatically restart itself and rebuild the WholeTale plugin
4. Navigate to SwaggerUI to find the new [`PUT /instance/:id` endpoint](https://girder.local.wholetale.org/api/v1#!/instance/instance_updateInstance)
5. Lookup the ID of an instance, enter it into the input field provided, and click "Try it out"
    * Your request should return 200 and the response is the instance you just PUT
6. Check `docker logs -f celery_worker`
    * You should see a line in the logs saying that the `update_container` task has succeeded
7. Check `docker ps` and/or `docker service ls`
    * ~~XXX: This is currently broken, due to a digest mismatch (see https://github.com/whole-tale/gwvolman/pull/44#issuecomment-454548260)~~
    * You should see that your instance has been recreated
    * NOTE: If the underlying image has changed, it may take one or more minutes to restart
